### PR TITLE
XWIKI-18811: Add autocomplete attribute to login form fields

### PR DESCRIPTION
**Pull request by Sereza7**
**Jira:** https://jira.xwiki.org/browse/XWIKI-18811

## PR Changes
* Added autocomplete to login form fields.
* Added autocomplete to registration form field, except for password where it was explicitely set to 'off'. See register_macros.vm

## View
No view change.

### CODE REVIEW

**Review by michitux — CHANGES_REQUESTED** [review]
Shouldn't you also change https://github.com/xwiki/xwiki-platform/blob/053d957d53f2a543d158f3ab651e390d2728e0b9/xwiki-platform-core/xwiki-platform-web/xwiki-platform-web-templates/src/main/resources/templates/registerinline.vm#L59-L99? Also, why don't you set `new-password` as `autocomplete` value for the new password? See, e.g., https://www.chromium.org/developers/design-documents/form-styles-that-chromium-understands/?

File: xwiki-platform-core/xwiki-platform-administration/xwiki-platform-administration-ui/src/main/resources/XWiki/Registration.xml [hunk 1]
```diff
@@ -457,8 +461,10 @@
         &lt;form id="loginForm" action="$loginURL" method="post"&gt;
           &lt;div class="centered"&gt;
           &lt;input type="hidden" name="form_token" value="$!{services.csrf.getToken()}" /&gt;
-          &lt;input id="j_username" name="j_username" type="hidden" value="$escapetool.xml($!userName)" /&gt;
-          &lt;input id="j_password" name="j_password" type="hidden" value="$escapetool.xml($!password)" /&gt;
+          &lt;input id="j_username" name="j_username" type="hidden" value="$escapetool.xml($!userName)"
+            autocomplete="username"/&gt;
+          &lt;input id="j_password" name="j_password" type="hidden" value="$escapetool.xml($!password)"
+            autocomplete="current-password"/&gt;
```
[the remark this answers is not in the document]
Comment (Sereza7) [reply xwiki-platform-core/xwiki-platform-administration/xwiki-platform-administration-ui/src/main/resources/XWiki/Registration.xml]: Those inputs are not shown at all. When successfully registering, the user is prompted with a login that submit an already filled form.
So no, there's no need to put an autocomplete attribute on those inputs.
Addressed in https://github.com/xwiki/xwiki-platform/pull/2195/commits/47f13dc69b8f0ef701a99bbe08517bb5974c3b0f

**Comment by michitux** [comment]
> So I think we should leave those fields on autocomplete='off'.

How do you reach that conclusion? The developer documentation of both Firefox and Chrome says that you should use `autocomplete="new-password"` for the password field in user sign-up forms.

For Firefox: https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/autocomplete:

> "new-password"
>    A new password. When creating a new account or changing passwords, this should be used for an "Enter your new password" or "Confirm new password" field, as opposed to a general "Enter your current password" field that might be present. This may be used by the browser both to avoid accidentally filling in an existing password and to offer assistance in creating a secure password (see also [Preventing autofilling with autocomplete="new-password"](https://developer.mozilla.org/en-US/docs/Web/Security/Securing_your_site/Turning_off_form_autocompletion#preventing_autofilling_with_autocompletenew-password)).

For Chrome: https://www.chromium.org/developers/design-documents/form-styles-that-chromium-understands/

> Add an [autocomplete](https://html.spec.whatwg.org/multipage/forms.html#autofilling-form-controls:-the-autocomplete-attribute) attribute with a value of [new-password](https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#attr-fe-autocomplete-new-password) for the password field on sign-up and change-password forms.

Also from Google Developers: https://web.dev/learn/forms/autofill/#ensure-users-can-sign-in-fast-and-use-secure-passwords

> As this is a sign-up form, users shouldn't get the option to fill in previously used passwords. You can use autocomplete="new-password" to ensure browsers offer the option to generate a new password.

Similarly, also Apple says you should use `new-password` when you expect the user to enter a new password: https://developer.apple.com/documentation/security/password_autofill/enabling_password_autofill_on_an_html_input_element

Therefore, all major browser vendors recommend using `autofill="new-password"` when you want the user to enter a new password during registration.

Further, browsers ignore `autocomplete="off"` on login fields, see https://developer.mozilla.org/en-US/docs/Web/Security/Securing_your_site/Turning_off_form_autocompletion#the_autocomplete_attribute_and_login_fields or also [this Stack Overflow answer](https://stackoverflow.com/a/28457066/1293930) - it is in fact by setting `autocomplete="new-password"` that you prevent auto-completion of the current password in the registration form.

**Comment by michitux** [comment]
> It's not supported by Chrome and Safari.

I doubt this information is accurate as both Chrome in the Chromium and Google developer documentation as well as the Apple developer documentation (that surely has Safari in mind) say the value should be used. See, e.g., [this article](https://www.amitmerchant.com/enabling-new-password-suggestions-in-password-input-in-html/) that clearly shows Chrome offering to create a new password when using this attribute.

**Comment by Sereza7** [comment]
@michitux 
> See, e.g., [this article](https://www.amitmerchant.com/enabling-new-password-suggestions-in-password-input-in-html/) that clearly shows Chrome offering to create a new password when using this attribute.

I tested in Chrome and could not find a difference in implementation between autocomplete='off' and autocomplete='new-password': here is stack overflow, using 'off', and getting autocompleted fine by Chrome.
image

> I doubt this information is accurate as both Chrome in the Chromium and Google developer documentation as well as the Apple developer documentation (that surely has Safari in mind) say the value should be used.

This information is probably imprecise, because there's no difference between using 'off' and 'new-password' for Chrome or Safari as of now. Seeing that most a lot of large websites still run with 'off', I'm in favor of leaving it like it is now in XWiki. If need be, it's an easy fix later, but I doubt this would be needed in the next 5 years at least.

**Comment by Sereza7** [comment]
Added the `new-password` value to the autocomplete attribute of both the password fields in the register form in https://github.com/xwiki/xwiki-platform/pull/2195/commits/2a68d7096252e50b13e3f1e53eebe61a38b85394.

___
This PR is ready for further reviews.

**Comment by Sereza7** [comment]
## PR Changes
* Fixed password fields in register_macros.vm and passwd.vm

___
Comment https://github.com/xwiki/xwiki-platform/pull/2195#pullrequestreview-1624430057 has been addressed in https://github.com/xwiki/xwiki-platform/pull/2195/commits/a6f3879dcb6a0fef3e08267dec6f0c8121f39875 :+1:

**Pull request merged**